iwyu.cc

```diff
@@ -454,32 +454,32 @@ class BaseAstVisitor : public RecursiveASTVisitor<Derived> {
       errs() << AnnotatedName(GetKindName(typeloc)) << PrintableTypeLoc(typeloc)
              << "\n";
     }
-    return Base::TraverseTypeLoc(typeloc);
+    return Base::TraverseTypeLoc(typeloc, traverse_qualifier);
   }
 
-  bool TraverseNestedNameSpecifier(NestedNameSpecifier* nns) {
+  bool TraverseNestedNameSpecifier(NestedNameSpecifier nns) {
     if (!nns)
       return true;
-    ASTNode node(nns);
+    ASTNode node(&nns);
```

Comment:
I'm a bit surprised the address of the argument `nns` is valid throughout the whole traversal of the subtree under `NestedNameSpecifier`. But it has to be, because `node` has the same lifetime, and that's definitely alive... Right?

Reply:
Yeah, correct. The same trick already was in `TraverseNestedNameSpecifierLoc` below.

Reply:
I think it's fine as-is. There doesn't appear to be a copy-ctor on NNS, so it takes some shuffling to store it by-val.

Reply:
Sorry? This compiles:
```cpp
NestedNameSpecifier nns2 = nns;
```
or I haven't understood you correctly.

Reply:
Oh, it does? I had problems earlier when I was playing with it, but I think I was targeting the wrong llvm tree or something. Anyway, I think the current impl is fine.